### PR TITLE
Ingest and mutation support for python sdk

### PR DESCRIPTION
**Pull request by mullr**

### CODE REVIEW

File: client-libraries/python/auxon-sdk/src/ingest.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    }
+}
+
+#[pymethods]
+impl IngestClient {
+    pub fn disable_auto_timestamp(&mut self) {
+        self.client.disable_auto_timestamp();
+    }
+
+    pub fn switch_timeline(&mut self, id: TimelineId) -> Result<(), SdkError> {
```
Comment (jonlamb-gh) [inline client-libraries/python/auxon-sdk/src/ingest.rs:29]: Does it matter that this returns SdkError while the others return PyErr?

File: client-libraries/python/auxon-sdk/src/ingest.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        name: &str,
+        timeline_attrs: &Bound<pyo3::types::PyDict>,
+    ) -> Result<(), PyErr> {
+        let attrs = py_dict_to_attr_vec(timeline_attrs)?;
+        self.rt
+            .block_on(
+                self.client
+                    .send_timeline_attrs(name, attrs.iter().map(|(k, v)| (k.as_str(), v.clone()))),
+            )
+            .map_err(SdkError::from)?;
```
Comment (jonlamb-gh) [inline client-libraries/python/auxon-sdk/src/ingest.rs:44]: I'm surprised we can't do the usual `impl From` and use implicit error type conversion in place of these `map_err`'s?

File: client-libraries/rust/examples/multi_mutator.rs [hunk 3]
```diff
@@ -28,6 +28,7 @@ impl BearingConfigFilePath for CliOptions {
 }
 
 fn main() {
+    #[allow(deprecated)]
```
Comment (jonlamb-gh) [inline client-libraries/rust/examples/multi_mutator.rs:31]: Would be nice to update these to use the new APIs

File: client-libraries/rust/src/plugin_utils/config.rs [hunk 4]
```diff
@@ -0,0 +1,936 @@
...
+    /// authenticate, and return a connected MutatorHost.
+    ///
+    /// This also creates an ingest connection, which is used
+    /// internally to log mutation-related events.
+    pub async fn connect_and_authenticate_mutation(
+        &self,
+    ) -> Result<super::mutation::MutatorHost, Box<dyn std::error::Error + Send + Sync>> {
+        let ingest = self.connect_and_authenticate_ingest().await?;
+
+        // TODO handle top level mutator_attributes configuration
```
Comment (jonlamb-gh) [inline client-libraries/rust/src/plugin_utils/config.rs:309]: Are we tracking this TODO somewhere?

File: client-libraries/rust/examples/utils_mutator.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    auxon_sdk::init_tracing!();
+
+    let value = Arc::new(AtomicI64::new(0));
+
+    println!("Connecting...");
+    let config = Config::load_common()?;
+    let mut mh = config.connect_and_authenticate_mutation().await?;
+    mh.register_mutator(Box::new(OffsetNumberMutator::new(value.clone())))
+        .await?;
+    // TODO allow mutators to be added after connect
```
Comment (jonlamb-gh) [inline client-libraries/rust/examples/utils_mutator.rs:25]: I don't see this scenario being explicitly rejected in the impl, is it actually unsupported?

File: client-libraries/python/auxon-sdk/tests/test_ingest.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+@dataclass
+class BaseConfig:
+    base_str_val: str = None
+    base_int_val: int = None
+    base_float_val: float = None
+    base_bool_val: bool = None
+
+@dataclass
+class SimpleConfig(BaseConfig):
```
Comment (jonlamb-gh) [inline client-libraries/python/auxon-sdk/tests/test_ingest.py:14]: Would be nice to see a test covering loading config from toml file too

**Review by jonlamb-gh — APPROVED** [review]
A few questions, and CI needs fixed up
but otherwise LGTM

File: client-libraries/python/auxon-sdk/src/ingest.rs (hunk 1 above)
Comment (mullr) [reply client-libraries/python/auxon-sdk/src/ingest.rs:29]: No, pyo3 does conversion automatically

File: client-libraries/python/auxon-sdk/src/ingest.rs (hunk 2 above)
Comment (mullr) [reply client-libraries/python/auxon-sdk/src/ingest.rs:44]: I tihnk we probably can in this case.

File: client-libraries/rust/examples/utils_mutator.rs (hunk 5 above)
Comment (mullr) [reply client-libraries/rust/examples/utils_mutator.rs:25]: It's a stale comment, I actually fixed this.

File: client-libraries/rust/src/plugin_utils/config.rs (hunk 4 above)
Comment (mullr) [reply client-libraries/rust/src/plugin_utils/config.rs:309]: Yeah, it's https://github.com/auxoncorp/modality/issues/2748

**Pull request merged**